Numbers concept (#1054)
exercism/haskell: concepts/numbers/introduction.md

# Haskell Built-in Numeric Types

> Integers` hold any number no matter how big, up to the limit of your machine's memory, while `Int` corresponds to the set of positive and negative integers that can be expressed in 32 or 64 bits (guaranteed at least -2^29 to 2^29).

Reviewer: If I look at the exemplar solution, it looks like only `Integer` is used, but not `Int`. In principle, this means that the section on `Int` should be removed, but I think in this particular case it is fine to keep it in.

`Float` corresponds to the set of real numbers, limited by the precision of the computer.

> Operations defined on numbers usually work on one type or the other, but not mixing them.

Reviewer: Maybe do:
```suggestion
Operations defined on numbers usually work on one type or the other, but not on both.
```

> Operations defined on numbers usually work on one type or the other, but not mixing them.

Reviewer: Maybe do:

```suggestion
Operations defined on numbers usually work on one type or the other, but not on both.
```

```haskell 
n :: Integer
n = 6
x :: Float
x = fromIntegral n      --> x = 6.0

m :: Int
m = 7
y :: Double

> y = fromIntegral m      --> y = 7.0

Author: Not sure I understood the intent so I removed the second example.

> Functions to convert between `Float` and `Integer`, include `toFloat` which converts `Int\Integer` to `Float` and `round`, `floor`, `ceiling` or `truncate` which convert a `Float` to an `Int`.

Reviewer: This is a fairly big list, most of which are functions that the student won't need. Normally, we don't list specific functions in the introduction, but only mention the package/module in which they can be found. This is to stimulate students to figure out finding these functions themselves, which will be great for their learning experience.

In this case, the above functions are all part of the prelude I think, correct? What I would suggest is to remove all but one or two functions.

> Conversion from an integer type (Int or Integer) to anything else is done using "fromIntegral".

Reviewer: ```suggestion
Conversion from an integer type (`Int` or `Integer`) to anything else is done using `fromIntegral`.
```

> ```

Reviewer: Excellent! Nice and concise.

# Introduction

> `Integers` hold any number no matter how big, up to the limit of your machine's memory, while `Int` corresponds to the set of positive and negative integers that can be expressed in 32 or 64 bits (guaranteed at least -2^29 to 2^29).

Reviewer: ```suggestion
`Integer`s hold any number no matter how big, up to the limit of your machine's memory, while `Int` corresponds to the set of positive and negative integers that can be expressed in 32 or 64 bits (guaranteed at least -2^29 to 2^29).
```

This comment also applies to the exercise introduction.

> # Haskell Built-in Numeric Types

Reviewer: There is quite some content in this file. While the `about.md` document can be quite elaborate, the `introduction.md` document should be quite succinct. See [the docs](https://exercism.org/docs/building/tracks/concept-exercises#h-file-docs-introduction-md) which have these rules (amongst others):

> The information provided should give the student just enough context to figure out the solution themselves.
>
> Only information that is needed to understand the fundamentals of the concept and solve the exercise should be provided. Extra information should be left for the concept's about.md document.

Haskell has a neat trick with operators: you can take any function that takes two arguments and use it like an operator by enclosing the name in back-ticks. So the following two lines mean exactly the same:

```haskell
 d = 7 `div` 3
 d = div 7 3
```

With that in mind, here are the integer division operators:

- `quot`: Returns the quotient of the two numbers. This is the result of division which is then truncated towards zero.

- `rem`: Returns the remainder from the quotient.

- `div`: Similar to "quot", but is rounded down towards minus infinity.

- `mod`: Returns the modulus of the two numbers. This is similar to the remainder, but has different rules when "div" returns a negative number.

Just as you can convert a function with two arguments into an operator, you can convert an operator into a function with two arguments by putting it in parentheses. So the following two lines mean the same thing:

```haskell
(+) 3 4
3 + 4
```

---
### Credits: 

> The above text is modified from __The Haskell 98 Report__ by Simon Peyton Jones, used with permission to copy, distribute and modify for any purpose with this note included. 

Reviewer: @SaschaMann Do you remember if we have any attribution guidelines?

Reviewer 2: (I'll unsubscribe from this issue. Please ping me in any responses to this comment chain in particular)
